cpu/rp2040/pll.c

```diff
@@ -0,0 +1,131 @@
+ */
+
+/**
+ * @ingroup     cpu_rp2040
+ * @{
+ *
+ * @file
+ * @brief       Implementation of the phase locked loop (PLL)
+ *              to drive the ROSC or XOSC to a higher frequency
+ *
+ * @author      Fabian Hüßler <fabian.huessler@ovgu.de>
+ *
+ * @}
+ */
+
+#include <assert.h>
+
+#include "macros/units.h"
+#include "periph_cpu.h"
+#include "vendor/RP2040.h"
+#include "io_reg.h"
+
+#define RP2040_REF_CLOCK_MIN        (MHZ(5U))
+#define RP2040_REF_CLOCK_MAX        (MHZ(15U))
+#define RP2040_REF_CLOCK_DIV_MIN    (1U)
+#define RP2040_REF_CLOCK_DIV_MAX    (63U)
+#define RP2040_VCO_FEEDBACK_DIV_MIN (16U)
+#define RP2040_VCO_FEEDBACK_DIV_MAX (320U)
+#define RP2040_POST_DIV_MIN         (1U)
+#define RP2040_POST_DIV_MAX         (7U)
+
+/**
+ * @brief   Start the PLL for the system clock (@p pll = PLL_SYS) or
+ *          the USB clock (@p pll = PLL_USB)
+ */
+static void _pll_start(PLL_SYS_Type *pll,
+                       uint32_t f_ref, uint8_t ref_div,
+                       uint16_t vco_feedback_scale,
+                       uint8_t post_div_1, uint8_t post_div_2)
+{
+    assert(RP2040_REF_CLOCK_MIN <= f_ref);
+    assert(f_ref <= RP2040_REF_CLOCK_MAX);
+    assert(RP2040_REF_CLOCK_DIV_MIN <= ref_div);
+    assert(ref_div <= RP2040_REF_CLOCK_DIV_MAX);
+    assert(RP2040_VCO_FEEDBACK_DIV_MIN <= vco_feedback_scale);
+    assert(vco_feedback_scale <= RP2040_VCO_FEEDBACK_DIV_MAX);
+    assert(RP2040_POST_DIV_MIN <= post_div_1);
+    assert(post_div_1 <= RP2040_POST_DIV_MAX);
+    assert(RP2040_POST_DIV_MIN <= post_div_2);
+    assert(post_div_2 <= RP2040_POST_DIV_MAX);
+
+    /* program reference clock divider */
+    io_reg_write_dont_corrupt(&pll->CS.reg, ref_div << PLL_SYS_CS_REFDIV_Pos,
+                              PLL_SYS_CS_REFDIV_Msk);
+    /* program feedback divider */
+    io_reg_write_dont_corrupt(&pll->FBDIV_INT.reg,
+                              vco_feedback_scale << PLL_SYS_FBDIV_INT_FBDIV_INT_Pos,
+                              PLL_SYS_FBDIV_INT_FBDIV_INT_Msk);
+    /* turn on the main power */
+    io_reg_atomic_clear(&pll->PWR.reg,
+                     (1U << PLL_SYS_PWR_VCOPD_Pos) |
+                     (1U << PLL_SYS_PWR_DSMPD_Pos) |
+                     (1U << PLL_SYS_PWR_PD_Pos));
+    /* wait for VCO to lock (i.e. keep its output stable) */
+    while (!pll->CS.bit.LOCK) { }
+    /* set up post divisors and turn them on */
+    pll->PRIM.reg = (post_div_1 << PLL_SYS_PRIM_POSTDIV1_Pos)
+                  | (post_div_2 << PLL_SYS_PRIM_POSTDIV2_Pos);
+    io_reg_atomic_clear(&pll->PWR.reg, 1U << PLL_SYS_PWR_POSTDIVPD_Pos);
+}
+
+/**
+ * @brief   Stop the PLL for the system clock (@p pll = PLL_SYS) or
+ *          the USB clock (@p pll = PLL_USB)
+ */
+static void _pll_stop(PLL_SYS_Type *pll)
+{
+    uint32_t reg = (1U << PLL_SYS_PWR_VCOPD_Pos) |
+                   (1U << PLL_SYS_PWR_POSTDIVPD_Pos) |
+                   (1U << PLL_SYS_PWR_DSMPD_Pos) |
+                   (1U << PLL_SYS_PWR_PD_Pos);
```

Comment:
IMHO

```suggestion
    uint32_t reg = (1U << PLL_SYS_PWR_VCOPD_Pos)
                 | (1U << PLL_SYS_PWR_POSTDIVPD_Pos)
                 | (1U << PLL_SYS_PWR_DSMPD_Pos)
                 | (1U << PLL_SYS_PWR_PD_Pos);
```


looks nicer